spring-boot-actuator/src/main/java/org/springframework/boot/actuate/health/DefaultHealthIndicatorRegistry.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2012-2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.boot.actuate.health;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import org.springframework.util.Assert;
+
+/**
+ * Default implementation of {@link HealthIndicatorRegistry}.
+ *
+ * @author Vedran Pavic
+ */
+public class DefaultHealthIndicatorRegistry implements HealthIndicatorRegistry {
+
+	private final ConcurrentMap<String, HealthIndicator> healthIndicators =
+			new ConcurrentHashMap<String, HealthIndicator>();
+
+	@Override
+	public void register(String name, HealthIndicator healthIndicator) {
+		Assert.notNull(healthIndicator, "HealthIndicator must not be null");
+		if (this.healthIndicators.putIfAbsent(name, healthIndicator) != null) {
+			throw new IllegalStateException(
+					"HealthIndicator with name '" + name + "' already registered");
+		}
+	}
+
+	@Override
+	public HealthIndicator unregister(String name) {
+		return this.healthIndicators.remove(name);
+	}
+
+	@Override
+	public HealthIndicator get(String name) {
+		return this.healthIndicators.get(name);
+	}
+
+	@Override
+	public Map<String, HealthIndicator> getAll() {
+		return Collections.unmodifiableMap(
+				new HashMap<String, HealthIndicator>(this.healthIndicators));
```

Comment:
Unfortunately, and somewhat surprisingly, this isn't thread-safe. It's doing a `putAll` under the covers, the documentation for which says:

> The behavior of this operation is undefined if the specified map is modified while the operation is in progress.

With the current implementation of `HashMap` it should work – it iterates over the concurrent hash map's entry set and doing so is thread-safe – but that's an implementation detail that we shouldn't rely on. As a result, everything needs to be `synchronized` which means there's no longer any point in using `ConcurrentHashMap`.